src/SIM/SIMinput.C

```diff
@@ -1072,12 +1076,105 @@ bool SIMinput::refine (const LR::RefineData& prm,
         return false;
       }
 
-      if (!pch->refine(prm,sol,fName))
+      if (myModel.size() > 1) {
+        LR::RefineData prmloc(prm);
+        if (prm.errors.size() > 0 ) { // refinement by true_beta
+          size_t n = myModel[i]->getNoElms();
+          prmloc.errors.resize(n);
+          for (size_t j = 0; j < n; j++) {
+            int el = myModel[i]->getElmID(j+1)-1;
+            prmloc.errors[j] = prm.errors[el];
+          }
+        } else {
+          prmloc.elements.clear();
+          for (auto &it : prm.elements) {
+            int node = myModel[i]->getNodeIndex(it+1)-1;
+            if(node > 0)
+              prmloc.elements.push_back(node);
+          }
+        }
+        if (!pch->refine(prmloc,sol,fName))
+          return false;
+      } else if (!pch->refine(prm,sol,fName))
         return false;
 
       isRefined = true;
     }
 
+  // fix up refinement for boundaries
+  if (isRefined && myModel.size() > 1)
+  {
+    bool extraRefine = true;
+    while (extraRefine)
+    {
+      extraRefine = false;
+      for (const auto& it : adm.dd.ghostConnections)
+      {
+        int sidx = this->getLocalPatchIndex(it.slave);
+        int midx = this->getLocalPatchIndex(it.master);
+
+        if (sidx > 0 && midx > 0) // TODO: mpi
+        {
+          ASMu2D* mpch2D = dynamic_cast<ASMu2D*>(this->getPatch(midx));
+          if (mpch2D) {
+            LR::parameterEdge sedge;
+            switch (it.sidx) {
+              case 1: sedge = LR::WEST; break;
+              case 2: sedge = LR::EAST; break;
+              default:
+              case  3: sedge = LR::SOUTH; break;
+              case  4: sedge = LR::NORTH; break;
+            }
+            LR::parameterEdge medge;
+            switch (it.midx) {
+              case 1: medge = LR::WEST; break;
+              case 2: medge = LR::EAST; break;
+              default:
+              case  3: medge = LR::SOUTH; break;
+              case  4: medge = LR::NORTH; break;
+            }
+            LR::LRSplineSurface *master = static_cast<ASMu2D*>(this->getPatch(midx))->getBasis(1);
+            LR::LRSplineSurface *slave  = static_cast<ASMu2D*>(this->getPatch(sidx))->getBasis(1);
+            extraRefine |= master->matchParametricEdge(medge, slave, sedge, it.orient);
+            master->enforceIsotropic();
+            slave->enforceIsotropic();
+            master->generateIDs();
+            slave->generateIDs();
```

Comment:
This (1120-1142) should be place in a method in ASMu2D instead, where *this is the master patch and the slave patch is provided as an ASMu2D argument together with the sidx and midx indices and the orient flag, to hide LR-specific stuff from the SIM-level. Similarly below for 3D.